File: src/main/java/com/rimlang/rim/lexer/Lexer.java
package com.rimlang.rim.lexer;

import com.rimlang.rim.util.TokenUtil;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class Lexer {

    private static int line = -1;
    private static int lineCharacter = -1;

    private static int i = 0;

    public static List<Token> lex(String code) {
        line = 0;
        lineCharacter = 0;
        i = 0;
        List<Token> tokens = new ArrayList<>();
        StringBuilder token = new StringBuilder();
        char[] charArray = code.toCharArray();
        while (i < charArray.length) {
            String string = String.valueOf(charArray[i]);
            if (string.matches("\\w")) {
                token.append(string);
            } else {
                if (token.length() > 0) {
                    String tokenString = token.toString();
                    if (tokenString.matches("-?\\d+(\\.\\d+)?")) {
                        tokens.add(new NumberToken(TokenType.NUMBER_LITERAL, Double.parseDouble(tokenString), line, lineCharacter));
                    } else {
                        TokenType type = TokenType.findType(tokenString); // Search for matching keywords
                        tokens.add(new StringToken(
                                Objects.requireNonNullElse(type, TokenType.ID),
                                tokenString,
                                line,
                                lineCharacter
                        ));
                    }
                    token = new StringBuilder();
                }
                if (string.equals("\n")) {
                    tokens.add(new StringToken(TokenType.EOL, "\\n", line, lineCharacter));
                    if (nextCharacter(charArray)) break;
                    continue;
                }
                if (string.equals("#")) {
                    // Begin comment
                    while (!string.equals("\n")) {
                        if (nextCharacter(charArray)) break;
                        string = String.valueOf(charArray[i]);
                    }
                    tokens.add(new StringToken(TokenType.EOL, "\\n", line, lineCharacter));
                    if (nextCharacter(charArray)) break;
                    continue;
                }
                if (i == charArray.length - 1) {
                    tokens.add(new StringToken(TokenType.findType(string), string, line, lineCharacter));
                } else {
                    String doubleToken = string + charArray[i + 1];
                    TokenType doubleTokenType = TokenType.findType(doubleToken);
                    if (doubleTokenType == null) {
                        tokens.add(new StringToken(TokenType.findType(string), string, line, lineCharacter));
                    } else {
                        tokens.add(new StringToken(doubleTokenType, doubleToken, line, lineCharacter));
                        if (nextCharacter(charArray)) break;
                        if (nextCharacter(charArray)) break;
                        continue;
                    }
                }
                if (string.equals("'") || string.equals("\"")) {
                    String stringType = string;
                    // Begin string
                    if (nextCharacter(charArray)) break;
                    string = String.valueOf(charArray[i]);
                    while (!string.equals(stringType)) {
                        token.append(string);
                        if (nextCharacter(charArray)) break;
                        string = String.valueOf(charArray[i]);
                    }
                    tokens.add(new StringToken(TokenType.STRING_LITERAL, token.toString(), line, lineCharacter));
                    token = new StringBuilder();
                }
            }
            if (nextCharacter(charArray)) break;
        }
        return groupBrackets(tokens
                .stream()
                .filter(t -> t.getType() != null)
                .collect(Collectors.toList()));
    }

    public static List<Token> groupBrackets(List<Token> code) {
        List<Token> groupedCode = new ArrayList<>();
        for (int j = 0; j < code.size(); j++) {
            Token token = code.get(j);
            Object value = token.value();
            if ("{".equals(value) || "[".equals(value) || "(".equals(value)) {
                int end = TokenUtil.next(code, j + 1, token.getType());
                Token group = new GroupToken(token.getType(),
                        groupBrackets(TokenUtil.sub(code, j + 1, end)),
                        token.getLine(),
                        token.getCharacter()
                );
                groupedCode.add(group);
                j = end;
                continue;
            }
            groupedCode.add(token);
        }
        return groupedCode;
    }

    private static boolean nextCharacter(char[] code) {
        if (i >= code.length - 1) return true;
        i++;
        lineCharacter++;
        if (code[i] == '\n') {
            line++;
            lineCharacter = 0;
        }
        return false;
    }

}

File: src/main/java/com/rimlang/rim/lexer/Lexer.kt
package com.rimlang.rim.lexer

import com.rimlang.rim.lexer.token.*
import com.rimlang.rim.lexer.token.GroupToken
import com.rimlang.rim.lexer.token.NumberToken
import com.rimlang.rim.lexer.token.StringToken
import com.rimlang.rim.errors.CodeFile
import com.rimlang.rim.lexer.token.nextToken
import com.rimlang.rim.lexer.token.sub

object Lexer {

    private var line = -1
    private var lineCharacter = -1
    private var i = 0

    fun lex(code: CodeFile): List<Token> {
        line = 0
        lineCharacter = 0
        i = 0
        val tokens = mutableListOf<Token>()
        var token = ""
        val charArray = code.text.toCharArray()
        while (i < charArray.size) {

            // Add to token
            var string = charArray[i].toString()
            if (string matches Regex("\\w")) {
                token += string
                if (nextCharacter(charArray)) break
                continue
            }

            // Identifier or number
            if (token.isNotEmpty()) {
                if (token matches Regex("-?\\d+(\\.\\d+)?")) {
                    tokens.add(NumberToken(TokenType.NUMBER_LITERAL, token.toDouble(), line, lineCharacter, code))
                } else {
                    val type = findTokenType(token) // Search for matching keywords
                    tokens.add(
                        StringToken(
                            type ?: TokenType.ID,
                            token,
                            line,
                            lineCharacter,
                            code
                        )
                    )
                }
                token = ""
            }

            // New line
            if (string == "\n") {
                tokens.add(StringToken(TokenType.EOL, "\\n", line, lineCharacter, code))
                if (nextCharacter(charArray)) break
                continue
            }

            // Comment
            if (string == "#") {
                while (string != "\n") {
                    if (nextCharacter(charArray)) break
                    string = charArray[i].toString()
                }
                tokens.add(StringToken(TokenType.EOL, "\\n", line, lineCharacter, code))
                if (nextCharacter(charArray)) break
                continue
            }

            // String literal
            if (string == "'" || string == "\"") {
                val stringType = string
                if (nextCharacter(charArray)) break
                string = charArray[i].toString()
                while (string != stringType) {
                    token += string
                    if (nextCharacter(charArray)) break
                    string = charArray[i].toString()
                }
                tokens.add(StringToken(TokenType.STRING_LITERAL, token, line, lineCharacter, code))
                token = ""
            }

            // Check for last character
            val type = findTokenType(string)
            if (i == charArray.size - 1) {
                if (type != null) {
                    tokens.add(StringToken(type, string, line, lineCharacter, code))
                }
                if (nextCharacter(charArray)) break
                continue
            }

            // Two character tokens
            val doubleToken = string + charArray[i + 1]
            val doubleTokenType = findTokenType(doubleToken)
            if (doubleTokenType == null) {
                if (type != null) {
                    tokens.add(StringToken(type, string, line, lineCharacter, code))
                }
                if (nextCharacter(charArray)) break
                continue
            }
            tokens.add(StringToken(doubleTokenType, doubleToken, line, lineCharacter, code))
            if (nextCharacter(charArray)) break
            if (nextCharacter(charArray)) break
        }
        return groupBrackets(tokens)
    }

    private fun groupBrackets(code: List<Token>): List<Token> {
        val groupedCode = mutableListOf<Token>()
        var j = 0
        while (j < code.size) {
            val token = code[j]
            val value = token.value
            if (value in listOf("{", "[", "(")) {
                val end = nextToken(code, j + 1, token.type)
                val group = GroupToken(
                    token.type,
                    groupBrackets(sub(code, j + 1, end)),
                    token.line,
                    token.character,
                    token.file
                )
                groupedCode.add(group)
                j = end + 1
                continue
            }
            groupedCode.add(token)
            j++
        }
        return groupedCode
    }

    private fun nextCharacter(code: CharArray): Boolean {
        if (i >= code.size - 1) return true
        i++
        lineCharacter++
        if (code[i] == '\n') {
            line++
            lineCharacter = 0
        }
        return false
    }
}